src/extension/runner.ts

```diff
@@ -0,0 +1,378 @@
+import {
+  CreateSessionRequest,
+  ExecuteRequest,
+  ExecuteResponse,
+  Session,
+} from '@buf/stateful_runme.community_timostamm-protobuf-ts/runme/runner/v1/runner_pb'
+import { ChannelCredentials } from '@grpc/grpc-js'
+import { GrpcTransport } from '@protobuf-ts/grpc-transport'
+import { DuplexStreamingCall } from '@protobuf-ts/runtime-rpc/build/types/duplex-streaming-call'
+import {
+  commands,
+  window,
+  ExtensionContext,
+  ExtensionTerminalOptions,
+  Pseudoterminal,
+  Event,
+  TerminalDimensions,
+  EventEmitter,
+  Disposable,
+} from 'vscode'
+
+import { RunnerServiceClient } from './grpc/client'
+import { UInt32Value } from '@buf/stateful_runme.community_timostamm-protobuf-ts/google/protobuf/wrappers_pb'
+import type { DisposableAsync } from '../types'
+
+type ExecuteDuplex = DuplexStreamingCall<ExecuteRequest, ExecuteResponse>
+
+export interface RunProgramOptions {
+  programName: string
+  args?: string[]
+  cwd?: string
+  envs?: string[]
+  script?: 
+    |{
+      type: 'commands',
+      commands: string[]
+    }
+    |{
+      type: 'script',
+      script: string
+    }
+  tty?: boolean
+  environment?: IRunnerEnvironment
+}
+
+export interface IRunner extends Disposable {
+  close(): void
+
+  createEnvironment(
+    envs?: string[],
+    metadata?: { [index: string]: string }
+  ): Promise<IRunnerEnvironment>
+
+  createProgramSession(opts: RunProgramOptions): Promise<IRunnerProgramSession>
+}
+
+export interface IRunnerEnvironment extends DisposableAsync { }
+
+export interface IRunnerProgramSession extends DisposableAsync, Pseudoterminal {
+  readonly onDidErr: Event<string>
+  readonly onDidClose: Event<void | number>;
+
+  readonly onStdoutRaw: Event<Uint8Array>
+  readonly onStderrRaw: Event<Uint8Array>
+
+  handleInput(message: string): Promise<void>
+
+  setRunOptions(opts: RunProgramOptions): void
+  run(): Promise<void>
+}
+
+// TODO: handle connection failures better!!
+export class GrpcRunner implements IRunner {
+  protected readonly client: RunnerServiceClient
+  protected transport: GrpcTransport
+
+  private children: WeakRef<DisposableAsync>[] = []
+
+  constructor() {
+    this.transport = new GrpcTransport({
+      host: 'unix:///tmp/runme/runme.sock',
```

Comment:
We are already using `unix:///tmp/runme.sock` and `localhost:8080`. let's not introduce a third option.

Reply:
Oh yeah that's my bad, will fix 👍